File: 926. Flip String to Monotone Increasing/Solution.java
class Solution {
    public int minFlipsMonoIncr(String s) {
        // return dp(s);
        return logic(s);
    }

    public int logic(String s) {
        int zeros = 0, oneToZeros = 0, n = s.length();
        for (int i = 0; i < n; i++) {
            if (s.charAt(i) == '0') {
                zeros++;
            } else {
                oneToZeros++;
            }
            zeros = Math.min(zeros, oneToZeros);
        }
        return zeros;
    }

    public int dp(String s) {
        int n = s.length(), res = 100001;
        int[] oneToZero = new int[n + 1], zeroToOne = new int[n + 1];
        for (int i = 0; i < n; i++) {
            oneToZero[i + 1] = (s.charAt(i) == '1') ? oneToZero[i] + 1 : oneToZero[i];
            zeroToOne[n - i - 1] = (s.charAt(n - i - 1) == '0') ? zeroToOne[n - i] + 1 : zeroToOne[n - i];
        }
        for (int i = 0; i < n; i++) {
            res = Math.min(res, oneToZero[i + 1] + zeroToOne[i]);
        }
        return res - 1;
    }
}

File: 926. Flip String to Monotone Increasing/Solution.py
class Solution:
    def minFlipsMonoIncr(self, s: str) -> int:
        # return self.dp(s)
        return self.logic(s)

    def logic(self, s):
        zeros, onesToZeros = 0, 0
        for c in s:
            if c == '0':
                zeros += 1
            else:
                onesToZeros += 1
            zeros = min(zeros, onesToZeros)
        return zeros

    def dp(self, s):
        n, res = len(s), 100001
        oneToZeros, zeroToOnes = [0 for i in range(n + 1)], [0 for i in range(n + 1)]
        for i in range(n):
            oneToZeros[i + 1] = oneToZeros[i] + 1 if s[i] == '1' else oneToZeros[i]
            zeroToOnes[n - i - 1] = zeroToOnes[n - i] + 1 if s[n - i - 1] == '0' else zeroToOnes[n - i]
        for i in range(n):
            res = min(res, oneToZeros[i + 1] + zeroToOnes[i])
        return res - 1